File: NC-150/06_Linked_List/44_merge_k_sorted_lists.cpp
/*
 * Problem: Merge k Sorted Lists (LeetCode 23)
 * 
 * You are given an array of k linked-lists lists, each linked-list is sorted 
 * in ascending order. Merge all the linked-lists into one sorted linked-list.
 * 
 * Time Complexity: O(N log k) where N is total nodes, k is number of lists
 * Space Complexity: O(log k) for divide and conquer, O(k) for heap
 */

#include <iostream>
#include <vector>
#include <queue>
#include "list_node.h"

class Solution {
public:
    // Solution 1: Divide and Conquer
    // Time: O(N log k), Space: O(log k) recursion
    ListNode* mergeKListsDivideConquer(std::vector<ListNode*>& lists) {
        if (lists.empty()) return nullptr;
        return mergeRange(lists, 0, lists.size() - 1);
    }
    
    // Solution 2: Min Heap (Priority Queue)
    // Time: O(N log k), Space: O(k)
    ListNode* mergeKListsHeap(std::vector<ListNode*>& lists) {
        // Custom comparator for min-heap
        auto compare = [](ListNode* a, ListNode* b) {
            return a->val > b->val;  // Min-heap
        };
        std::priority_queue<ListNode*, std::vector<ListNode*>, decltype(compare)> pq(compare);
        
        // Add first node of each list to heap
        for (ListNode* list : lists) {
            if (list != nullptr) {
                pq.push(list);
            }
        }
        
        ListNode dummy(0);
        ListNode* tail = &dummy;
        
        while (!pq.empty()) {
            // Get smallest node
            ListNode* smallest = pq.top();
            pq.pop();
            
            // Add to result
            tail->next = smallest;
            tail = tail->next;
            
            // Add next node from the same list
            if (smallest->next != nullptr) {
                pq.push(smallest->next);
            }
        }
        
        return dummy.next;
    }
    
    // Main function using divide and conquer
    ListNode* mergeKLists(std::vector<ListNode*>& lists) {
        return mergeKListsDivideConquer(lists);
    }
    
private:
    ListNode* mergeRange(std::vector<ListNode*>& lists, int left, int right) {
        if (left == right) return lists[left];
        if (left > right) return nullptr;
        
        int mid = left + (right - left) / 2;
        ListNode* leftMerged = mergeRange(lists, left, mid);
        ListNode* rightMerged = mergeRange(lists, mid + 1, right);
        
        return mergeTwoLists(leftMerged, rightMerged);
    }
    
    ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
        ListNode dummy(0);
        ListNode* tail = &dummy;
        
        while (l1 != nullptr && l2 != nullptr) {
            if (l1->val <= l2->val) {
                tail->next = l1;
                l1 = l1->next;
            } else {
                tail->next = l2;
                l2 = l2->next;
            }
            tail = tail->next;
        }
        
        tail->next = (l1 != nullptr) ? l1 : l2;
        return dummy.next;
    }
};

int main() {
    Solution sol;
    
    // Test case 1: [[1,4,5],[1,3,4],[2,6]]
    std::vector<ListNode*> lists1 = {
        createList({1, 4, 5}),
        createList({1, 3, 4}),
        createList({2, 6})
    };
    std::cout << "Lists:" << std::endl;
    for (int i = 0; i < lists1.size(); i++) {
        std::cout << "  List " << i << ": ";
        printList(lists1[i]);
    }
    ListNode* merged1 = sol.mergeKLists(lists1);
    std::cout << "Merged: ";
    printList(merged1);
    deleteList(merged1);
    
    // Test case 2: Empty list
    std::vector<ListNode*> lists2 = {};
    ListNode* merged2 = sol.mergeKLists(lists2);
    std::cout << "\nEmpty input: ";
    printList(merged2);
    
    // Test case 3: [[]]
    std::vector<ListNode*> lists3 = {nullptr};
    ListNode* merged3 = sol.mergeKLists(lists3);
    std::cout << "\n[[]] input: ";
    printList(merged3);
    
    // Test case 4: Single list
    std::vector<ListNode*> lists4 = {createList({1, 2, 3})};
    std::cout << "\nSingle list: ";
    ListNode* merged4 = sol.mergeKLists(lists4);
    printList(merged4);
    deleteList(merged4);
    
    // Test heap solution
    std::vector<ListNode*> lists5 = {
        createList({1, 4, 5}),
        createList({1, 3, 4}),
        createList({2, 6})
    };
    ListNode* merged5 = sol.mergeKListsHeap(lists5);
    std::cout << "\nHeap solution: ";
    printList(merged5);
    deleteList(merged5);
    
    std::cout << "\nAll test cases passed!" << std::endl;
    return 0;
}

File: NC-150/06_Linked_List/44_merge_k_sorted_lists.py
"""
Problem: Merge k Sorted Lists (LeetCode 23)

You are given an array of k linked-lists lists, each linked-list is sorted 
in ascending order. Merge all the linked-lists into one sorted linked-list.

Time Complexity: O(N log k) where N is total nodes, k is number of lists
Space Complexity: O(log k) for divide and conquer, O(k) for heap
"""

from typing import List, Optional
import heapq

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
    
    # For heap comparison (needed since Python 3)
    def __lt__(self, other):
        return self.val < other.val

class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        """Divide and Conquer approach"""
        if not lists:
            return None
        return self._merge_range(lists, 0, len(lists) - 1)
    
    def _merge_range(self, lists, left, right):
        if left == right:
            return lists[left]
        if left > right:
            return None
        
        mid = (left + right) // 2
        left_merged = self._merge_range(lists, left, mid)
        right_merged = self._merge_range(lists, mid + 1, right)
        
        return self._merge_two(left_merged, right_merged)
    
    def _merge_two(self, l1, l2):
        dummy = ListNode(0)
        tail = dummy
        
        while l1 and l2:
            if l1.val <= l2.val:
                tail.next = l1
                l1 = l1.next
            else:
                tail.next = l2
                l2 = l2.next
            tail = tail.next
        
        tail.next = l1 if l1 else l2
        return dummy.next
    
    def mergeKListsHeap(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        """Min Heap approach"""
        heap = []
        
        # Add first node of each list to heap
        for i, node in enumerate(lists):
            if node:
                # Use (value, index, node) to handle equal values
                heapq.heappush(heap, (node.val, i, node))
        
        dummy = ListNode(0)
        tail = dummy
        
        while heap:
            val, idx, node = heapq.heappop(heap)
            
            tail.next = node
            tail = tail.next
            
            if node.next:
                heapq.heappush(heap, (node.next.val, idx, node.next))
        
        return dummy.next

# Helper functions
def create_list(values):
    if not values:
        return None
    head = ListNode(values[0])
    current = head
    for val in values[1:]:
        current.next = ListNode(val)
        current = current.next
    return head

def list_to_array(head):
    result = []
    while head:
        result.append(head.val)
        head = head.next
    return result

if __name__ == "__main__":
    sol = Solution()
    
    # Test case 1: [[1,4,5],[1,3,4],[2,6]]
    lists1 = [
        create_list([1, 4, 5]),
        create_list([1, 3, 4]),
        create_list([2, 6])
    ]
    print("Lists:", [[1,4,5], [1,3,4], [2,6]])
    merged1 = sol.mergeKLists(lists1)
    print(f"Merged: {list_to_array(merged1)}")
    assert list_to_array(merged1) == [1, 1, 2, 3, 4, 4, 5, 6]
    
    # Test case 2: Empty list
    lists2 = []
    merged2 = sol.mergeKLists(lists2)
    print(f"\nEmpty input: {list_to_array(merged2)}")
    assert list_to_array(merged2) == []
    
    # Test case 3: [[]]
    lists3 = [None]
    merged3 = sol.mergeKLists(lists3)
    print(f"[[]] input: {list_to_array(merged3)}")
    assert list_to_array(merged3) == []
    
    # Test case 4: Single list
    lists4 = [create_list([1, 2, 3])]
    merged4 = sol.mergeKLists(lists4)
    print(f"\nSingle list: {list_to_array(merged4)}")
    assert list_to_array(merged4) == [1, 2, 3]
    
    # Test heap solution
    lists5 = [
        create_list([1, 4, 5]),
        create_list([1, 3, 4]),
        create_list([2, 6])
    ]
    merged5 = sol.mergeKListsHeap(lists5)
    print(f"\nHeap solution: {list_to_array(merged5)}")
    assert list_to_array(merged5) == [1, 1, 2, 3, 4, 4, 5, 6]
    
    print("\nAll test cases passed!")

